docs/rfcs/014-storage-lsm.md

```diff
@@ -0,0 +1,142 @@
+# Why LSM trees?
+
+In general, an LSM tree has the nice property that random updates are
+fast, but the disk writes are sequential. When a new file is created,
+it is immutable. New files are created and old ones are deleted, but
+existing files are never modified. That fits well with storing the
+files on S3.
+
+Currently, we create a lot of small files. That is mostly a problem
+with S3, because each GET/PUT operation is expensive, and LIST
+operation only returns 1000 objects at a time, and isn't free
+either. Currently, the files are "archived" together into larger
+checkpoint files before they're uploaded to S3 to alleviate that
+problem, but garbage collecting data from the archive files would be
+difficult and we have not implemented it. This proposal addresses that
+problem.
+
+
+# Overview
+
+
+```
+^ LSN
+|
+|      Memtable:     +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|
+|            L0:     +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|
+|           L1:      +-------+ +-----+ +--+  +-+
+|                    |       | |     | |  |  | |
+|                    |       | |     | |  |  | |
+|                    +-------+ +-----+ +--+  +-+
+|
+|                       +----+ +-----+ +--+  +----+
+|                       |    | |     | |  |  |    |
+|                       |    | |     | |  |  |    |
+|                       +----+ +-----+ +--+  +----+
+|
++--------------------------------------------------------------> Page ID
+
+
++---+
+|   |   Layer file
++---+
+```
+
+
+# Memtable
+
+When new WAL arrives, it is first put into the Memtable. Despite the
+name, the Memtable is not a purely in-memory data structure. It can
+spill to a temporary file on disk if the system is low on memory, and
+is accessed through a buffer cache.
+
+If the page server crashes, the Memtable is lost. It is rebuilt by
+processing again the WAL that's newer than the latest layer in L0.
+
+The size of the Memtable is equal to the "checkpoint distance", or the
+amount of WAL that we need to keep in the safekeeper.
```

Comment:
> What happens if checkpointer is unable to keep up with the load?

The flushing of the in-memory layer is now done by separate "layer flushing thread". That makes it more likely that it can keep up, as the thread doesn't have any other duties. (No guarantee, though)

> Will memtable grow until OOM or the load will be throttled?

It will accumulate frozen in-memory layers, each the size of 'checkpoint_distance', until OOM. There's no throttling.

The in-memory layers are spilled to disk, using the ephemeral files, so it won't lead to OOM as quickly as you might think. However, the hashmap of page versions that's part of each in-memory layers is not spilled to disk, only the actual WAL records. So even if a layer is fully spilled to disk, it still consumes maybe 10% of that size of RAM.

Eventually, the compute node backpressure will kick in and throttle. But there's no throttling in the page server, and if you have a lot of tenants and all of them are being overloaded, you could still OOM.

I opened https://github.com/zenithdb/zenith/issues/1411 to track that.

Reply:
> > the amount of WAL that we need to keep in the safekeeper
> 
> I don't get this part of the sentence. Do we _need_ to keep anything in the safekeeper?

I rephrased this to:
```
+The size of the Memtable is configured by the "checkpoint distance"
+setting. Because anything that hasn't been flushed to disk and
+uploaded to S3 yet needs to be kept in the safekeeper, the "checkpoint
+distance" also determines the amount of WAL that needs to kept in the
+safekeeper.
```
Is that better?